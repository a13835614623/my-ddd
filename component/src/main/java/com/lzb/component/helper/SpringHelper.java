package com.lzb.component.helper;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.context.EnvironmentAware;
import org.springframework.core.env.Environment;
import org.springframework.core.env.StandardEnvironment;
import org.springframework.stereotype.Component;

@Component
public class SpringHelper implements ApplicationContextAware, EnvironmentAware {

    private static ApplicationContext APPLICATION_CONTEXT;
    private static StandardEnvironment ENVIRONMENT;
    private final static Pattern PATTERN = Pattern.compile("(?<=\\$\\{)[A-Za-z_\\-0-9.]+");
    @Override
    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
        SpringHelper.APPLICATION_CONTEXT = applicationContext;
    }

    @Override
    public void setEnvironment(Environment environment) {
        SpringHelper.ENVIRONMENT = (StandardEnvironment) environment;
    }

    private static ApplicationContext getApplicationContext() {
        return APPLICATION_CONTEXT;
    }

    public static Object getBean(String name) {
        return getApplicationContext().getBean(name);
    }

    public static <T> T getBean(Class<T> clazz) {
        return getApplicationContext().getBean(clazz);
    }

    public static <T> T getBean(String name, Class<T> clazz) {
        return getApplicationContext().getBean(name, clazz);
    }

    public static String getProperty(String propertyName, String defaultValue) {
        return ENVIRONMENT.getProperty(real(propertyName), defaultValue);
    }
    public <T> T getProperty(String propertyName, Class<T> targetType, T defaultValue) {
        return ENVIRONMENT.getProperty(real(propertyName), targetType, defaultValue);
    }

    private static String real(String propertyName) {
        if (propertyName.startsWith("${")) {
            Matcher matcher = PATTERN.matcher(propertyName);
            if (matcher.find()) {
                return matcher.group();
            }
        }
        return propertyName;
    }

}