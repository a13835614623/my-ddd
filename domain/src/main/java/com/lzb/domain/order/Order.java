package com.lzb.domain.order;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Objects;

import cn.hutool.core.lang.UUID;
import com.lzb.domain.common.BaseAggregate;
import com.lzb.domain.order.event.OrderCanceledEvent;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
// builder 反序列化
@SuperBuilder
@Jacksonized
public class Order extends BaseAggregate<Order> {

    @NonNull
    private String orderStatus;

    @NonNull
    private String currency;

    @NonNull
    private BigDecimal exchangeRate;

    @NonNull
    private BigDecimal totalShouldPay;

    @NonNull
    private BigDecimal totalActualPay;

    @NonNull
    private OrderAddress orderAddress;

    @Builder.Default
    private OrderDetails orderDetails = new OrderDetails(new ArrayList<>());

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = "CANCEL";
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(
                UUID.randomUUID().toString(),
                Objects.toString(id),
                Objects.toString(id),
                id));
    }
}
