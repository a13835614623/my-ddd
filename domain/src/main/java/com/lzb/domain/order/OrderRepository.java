package com.lzb.domain.order;

import com.lzb.domain.common.repository.CommonRepository;

/**
 * <br/>
 * Created on : 2023-08-30 23:07
 * @author mac
 */
public interface OrderRepository extends CommonRepository<Order> {

}
