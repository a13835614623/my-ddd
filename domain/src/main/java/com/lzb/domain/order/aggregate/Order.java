package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.exception.BizException;
import com.lzb.domain.common.aggregate.BaseAggregate;
import com.lzb.domain.order.aggregate.builder.OrderAddressBuilder;
import com.lzb.domain.order.dto.LockStockDto;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.domain.order.event.OrderCanceledEvent;
import com.lzb.domain.order.event.OrderPlacedEvent;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.Getter;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
public class Order extends BaseAggregate<Order> {

    @NonNull
    private final OrderStatus orderStatus;

    @NonNull
    private final String currency;

    @NonNull
    private final BigDecimal exchangeRate;

    @NonNull
    private final BigDecimal totalShouldPay;

    @NonNull
    private final BigDecimal totalActualPay;

    @NonNull
    private final OrderAddress orderAddress;

    @NonNull
    private final OrderDetails orderDetails;


    /**
     * 仓储层使用，无需校验业务逻辑，直接构造
     * @param id
     * @param version
     * @param orderStatus
     * @param currency
     * @param exchangeRate
     * @param totalShouldPay
     * @param totalActualPay
     * @param orderAddress
     * @param orderDetails
     */
    public Order(long id,
            int version,
            @NonNull OrderStatus orderStatus,
            @NonNull String currency,
            @NonNull BigDecimal exchangeRate,
            @NonNull BigDecimal totalShouldPay,
            @NonNull BigDecimal totalActualPay,
            @NonNull OrderAddress orderAddress,
            @NonNull List<OrderDetail> orderDetails) {
        super(id);
        this.version = version;
        this.orderStatus = orderStatus;
        this.currency = currency;
        this.exchangeRate = exchangeRate;
        this.totalShouldPay = totalShouldPay;
        this.totalActualPay = totalActualPay;
        this.orderAddress = orderAddress;
        this.orderDetails = new OrderDetails(orderDetails);
    }


    public void updateAddress(String email,
            String phoneNumber, FullName fullName,
            FullAddressLine fullAddressLine, String country) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        OrderAddress newOrderAddress = OrderAddressBuilder.newInstance(this.orderAddress)
                .addressLine1(fullAddressLine.getAddressLine1())
                .addressLine2(fullAddressLine.getAddressLine2())
                .country(country)
                .email(email)
                .firstName(fullName.getFirstName())
                .lastName(fullName.getLastName())
                .phoneNumber(phoneNumber).build();
        this.orderAddress = newOrderAddress;
    }

    /**
     * 更新库存结果
     * @param lockStock
     */
    public void updateLockStock(LockStockDto lockStock) {
        Map<Integer, List<OrderDetail>> skuId2OrderDetails = StreamEx.of(orderDetails.toStream()).groupingBy(OrderDetail::getSkuId);
        skuId2OrderDetails.keySet().forEach(skuId -> {
            List<OrderDetail> skuOrderDetails = skuId2OrderDetails.getOrDefault(skuId, Collections.emptyList());
            for (int i = 1; i <= skuOrderDetails.size(); i++) {
                skuOrderDetails.get(i - 1).updateLocked(i <= lockStock.getDetail(skuId).getLockedNum());
            }
        });
    }

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = OrderStatus.CANCELED;
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return OrderStatus.CANCELED == orderStatus;
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 只更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        this.orderAddress.updateFullName(fullName);
    }
}
