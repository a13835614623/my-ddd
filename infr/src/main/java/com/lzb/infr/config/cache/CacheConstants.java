package com.lzb.infr.config.cache;

import lombok.experimental.UtilityClass;

/**
 * <br/>
 * Created on : 2023-09-03 12:20
 * @author mac
 */
@UtilityClass
public final class CacheConstants {

    public static final String ORDER = "order";

}
