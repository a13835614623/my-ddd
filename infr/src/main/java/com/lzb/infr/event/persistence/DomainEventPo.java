package com.lzb.infr.event.persistence;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import com.lzb.infr.common.BasePo;
import com.lzb.infr.event.typehandler.ContentTypeHandler;
import lombok.Data;
import lombok.EqualsAndHashCode;

/**
 * <p>
 * 
 * </p>
 *
 * @author lizebin
 * @since 2023-09-02
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("domain_event")
public class DomainEventPo extends BasePo {

    private String topic;

    private String tag;

    private String bizId;

    private String key;

    @TableField(typeHandler = ContentTypeHandler.class)
    private String content;

    private Boolean sent;

    private String msgId;

}