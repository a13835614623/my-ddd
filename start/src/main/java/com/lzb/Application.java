package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@SpringBootApplication
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
