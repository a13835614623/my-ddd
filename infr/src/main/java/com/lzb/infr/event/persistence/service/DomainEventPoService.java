package com.lzb.infr.event.persistence.service;

import com.baomidou.mybatisplus.extension.service.IService;
import com.baomidou.mybatisplus.extension.service.impl.ServiceImpl;
import com.lzb.infr.event.persistence.DomainEventPo;
import com.lzb.infr.event.persistence.mapper.DomainEventPoMapper;

import org.springframework.stereotype.Service;

/**
 * <p>
 *  服务实现类
 * </p>
 *
 * @author lizebin
 * @since 2023-09-02
 */
@Service
public class DomainEventPoService extends ServiceImpl<DomainEventPoMapper, DomainEventPo> implements IService<DomainEventPo> {

}