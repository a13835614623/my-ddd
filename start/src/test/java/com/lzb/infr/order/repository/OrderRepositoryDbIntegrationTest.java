package com.lzb.infr.order.repository;

import java.math.BigDecimal;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneId;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.Resource;

import cn.hutool.core.util.IdUtil;
import com.lzb.BaseIntegrationTest;
import com.lzb.component.helper.SpringHelper;
import com.lzb.domain.order.aggregate.Order;
import com.lzb.domain.order.repository.OrderRepository;
import com.lzb.infr.config.cache.CacheConstants;
import com.lzb.infr.domain.order.repository.OrderRepositoryDb;
import com.lzb.infr.event.persistence.DomainEventPo;
import com.lzb.infr.event.persistence.service.DomainEventPoService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;

import org.springframework.cache.CacheManager;
import org.springframework.test.context.jdbc.Sql;

import static java.time.Instant.ofEpochMilli;

class OrderRepositoryDbIntegrationTest extends BaseIntegrationTest {

    @Resource(name = OrderRepositoryDb.BEAN_NAME)
    private OrderRepository orderRepository;

    @Resource
    private DomainEventPoService domainEventPoService;

    @Resource
    private CacheManager cacheManager;

    @Resource
    private SpringHelper springHelper;

    @Test
    @DisplayName("测试SpringHelper")
    void should_get_bean_from_spring_helper() {
        assertThat(springHelper.getBean(OrderRepositoryDb.BEAN_NAME)).isNotNull();
    }

    @Test
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_order_get.sql")
    @DisplayName("测试聚合根查询")
    void should_order_get() {

        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        assertThat(order.snapshot()).isNotNull();

        Map<String, Order> excepted = new HashMap<>();
        excepted.put("current", order);
        excepted.put("snapshot", order.snapshot());

        assertJSON(excepted);
    }

    @Test
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_update_order_info.sql")
    @DisplayName("测试聚合根查询")
    void should_update_order_info() {
        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        order.updateTotalActualPay(new BigDecimal("8888"));
        orderRepository.update(order);
        Order newOrder = orderRepository.get(orderId).orElseThrow();
        assertJSON(newOrder);
    }

    @Test
    @DisplayName("测试取消订单")
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_cancel_order.sql")
    void should_cancel_order() {
        // given
        long orderId = 1L;
        Order order = orderRepository.get(orderId).orElseThrow();
        Clock constantClock = Clock.fixed(ofEpochMilli(0), ZoneId.systemDefault());

        // DomainEvent设置了默认值，这里设置固定返回
        String msgId = "1";
        try (
            MockedStatic<Instant> instant = mockStatic(Instant.class);
            MockedStatic<IdUtil> idUtil = mockStatic(IdUtil.class);
        ) {
            instant.when(() -> Instant.now()).thenReturn(constantClock.instant());
            idUtil.when(() -> IdUtil.randomUUID()).thenReturn(msgId);
            order.cancel();
        }

        // when
        orderRepository.update(order);

        // then
        Order newOrder = orderRepository.get(orderId).orElseThrow();
        List<DomainEventPo> events = domainEventPoService.list();
        assertThat(events).hasSize(1);

        var assertMap = new HashMap<>();
        assertMap.put("event", events.get(0).getContent());
        assertMap.put("order", newOrder);
        assertJSON(assertMap);
    }

    @Test
    @DisplayName("测试缓存查询")
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_get_in_cache.sql")
    void should_get_in_cache() {

        long orderId = 1L;
        Order order1 = orderRepository.getInCache(orderId);
        Order order2 = orderRepository.getInCache(orderId);

        verify(cacheManager, atLeast(1)).getCache(CacheConstants.ORDER);
        assertThat(order1.snapshot()).isNull();
        assertThat(order2.snapshot()).isNull();
        var assertMap = Map.of("order1", order1, "order2", order2);
        assertJSON(assertMap);

    }

    @Test
    @DisplayName("测试缓存清除")
    @Sql("/sql/OrderRepositoryDbImplIntegrationTest/should_clear_cache_when_get_in_cache.sql")
    void should_clear_cache_when_get_in_cache() {
        long orderId = 1L;
        Order order = orderRepository.getInCache(orderId);
        assertThat(order).isNotNull();

        order = orderRepository.getOrThrow(orderId);
        order.cancel();
        orderRepository.update(order);

        Order cacheOrder = orderRepository.getInCache(orderId);
        assertThat(cacheOrder.isCancel()).isTrue();
        assertThat(cacheOrder.getVersion()).isEqualTo(2);
        assertJSON(cacheOrder);
    }

}