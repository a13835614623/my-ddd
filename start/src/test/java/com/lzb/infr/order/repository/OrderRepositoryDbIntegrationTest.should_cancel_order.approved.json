{
  "event" : "{\"key\": \"1\", \"tag\": \"ORDER_CANCELED\", \"bizId\": \"1\", \"msgId\": \"1\", \"orderId\": 1, \"timestamp\": 0}",
  "order" : {
    "currency" : "CNY",
    "events" : [ ],
    "exchangeRate" : 1,
    "id" : 1,
    "logs" : [ ],
    "orderAddress" : {
      "country" : "CHINA",
      "email" : "123456@qq.com",
      "fullAddressLine" : {
        "addressLine1" : "line1",
        "addressLine2" : "line2"
      },
      "fullName" : {
        "firstName" : "first",
        "lastName" : "last"
      },
      "id" : 1,
      "phoneNumber" : "13800000000",
      "version" : 1
    },
    "orderDetails" : [ {
      "id" : 1,
      "orderId" : 1,
      "orderStatus" : "CANCELED",
      "price" : 1,
      "skuId" : 1,
      "version" : 1
    }, {
      "id" : 2,
      "orderId" : 1,
      "orderStatus" : "CANCELED",
      "price" : 10,
      "skuId" : 2,
      "version" : 1
    } ],
    "orderStatus" : "CANCELED",
    "totalActualPay" : 88,
    "totalShouldPay" : 88,
    "version" : 2
  }
}